File: frontend/src/contracts/historian.ts
import type { Alert } from "./alerts";
import type { Prediction } from "./predictions";

export type HistorianFactCategory = "diagnosis" | "medication" | "lab" | "icu_event";
export type ContextualStatus = "complete" | "incomplete";

export interface HistorianFact {
  fact_id: string;
  category: HistorianFactCategory;
  label: string;
  value: string | null;
  unit: string | null;
  effective_at: string;
  source_kind: "synthetic";
  source_name: string;
}

export interface RuleEvaluation {
  rule_key: string;
  rule_name: string;
  rule_version: string;
  category: HistorianFactCategory;
  fact_id: string;
  delta: number;
  explanation: string;
  evaluated_at: string;
}

export interface TimelineEntry {
  entry_id: string;
  entry_type: "fact" | "prediction" | "alert" | "audit" | "annotation";
  occurred_at: string;
  title: string;
  detail: string;
}

export interface Annotation {
  annotation_id: number;
  author_id: string;
  text: string;
  created_at: string;
  source_label: string;
}

export interface HistorianResponse {
  patient_id: string;
  patient_name: string;
  admission_id: string;
  admitted_at: string;
  bed_id: string;
  unit: string;
  current_prediction: Prediction & {
    timestamp: string;
    current_vitals: Record<string, unknown>;
    provenance: Record<string, unknown>;
    prototype_label: string;
    contract_version: string;
    source_kind: "ml" | "deterministic_fallback";
    source_version: string;
    fallback_reason: string | null;
  };
  baseline_score: number;
  contextual_status: ContextualStatus;
  contextual_score: number | null;
  facts: HistorianFact[];
  rule_evaluations: RuleEvaluation[];
  missing_evidence: string[];
  annotations: Annotation[];
  alert: Alert | null;
  timeline: TimelineEntry[];
  prototype_label: string;
  provenance: "synthetic" | "research-prototype";
}

export interface AnnotationCreate {
  text: string;
}

File: backend/app/contracts/historian.py
from datetime import datetime
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field

from backend.app.contracts.alerts import AlertResponse
from backend.app.contracts.predictions import PredictionResponse


class HistorianFact(BaseModel):
    model_config = ConfigDict(extra="forbid", from_attributes=True)
    fact_id: str
    category: Literal["diagnosis", "medication", "lab", "icu_event"]
    label: str
    value: str | None
    unit: str | None
    effective_at: datetime
    source_kind: Literal["synthetic"]
    source_name: str


class RuleEvaluation(BaseModel):
    model_config = ConfigDict(extra="forbid")
    rule_key: str
    rule_name: str
    rule_version: str
    category: Literal["diagnosis", "medication", "lab", "icu_event"]
    fact_id: str
    delta: float = Field(ge=-1, le=1)
    explanation: str
    evaluated_at: datetime


class TimelineEntry(BaseModel):
    model_config = ConfigDict(extra="forbid")
    entry_id: str
    entry_type: Literal["fact", "prediction", "alert", "audit", "annotation"]
    occurred_at: datetime
    title: str
    detail: str


class AnnotationCreate(BaseModel):
    model_config = ConfigDict(extra="forbid")
    text: str = Field(min_length=1, max_length=500)


class AnnotationResponse(BaseModel):
    model_config = ConfigDict(extra="forbid", from_attributes=True)
    annotation_id: int
    author_id: str
    text: str
    created_at: datetime
    source_label: str


class HistorianResponse(BaseModel):
    model_config = ConfigDict(extra="forbid")
    patient_id: str
    patient_name: str
    admission_id: str
    admitted_at: datetime
    bed_id: str
    unit: str
    current_prediction: PredictionResponse
    baseline_score: float = Field(ge=0, le=1)
    contextual_status: Literal["complete", "incomplete"]
    contextual_score: float | None = Field(default=None, ge=0, le=1)
    facts: list[HistorianFact]
    rule_evaluations: list[RuleEvaluation]
    missing_evidence: list[str]
    annotations: list[AnnotationResponse]
    alert: AlertResponse | None
    timeline: list[TimelineEntry]
    prototype_label: str
    provenance: Literal["synthetic", "research-prototype"]